File: src_matlab/+nmpccodegen/+controller/+obstacles/Nonconvex_constraints.m
classdef Nonconvex_constraints < nmpccodegen.controller.obstacles.Obstacle
    %NON_CONVEX_CONSTRAINTS obstacle excisting of non convex
    %constrains
    %   This obstacle allows the users to easily find a custom object.
    
    properties(Access = private)
        constraints
    end
    methods
        function obj = Nonconvex_constraints(model)
            % - model = Model of the controlled system.
            obj@nmpccodegen.controller.obstacles.Obstacle(model)
            obj.constraints={};
        end
        function obj = add_constraint(obj,constraint)
            % Add an nonconvex constraint.
            %   constraint = Matlab function.
            
            number_of_constraints = obj.get_number_of_constraints();
            if(number_of_constraints==0)
                obj.constraints ={constraint};
            else
                obj.constraints{end+1} = constraint;
            end
        end
        function number_of_constraints = get_number_of_constraints(obj)
            % Get the number of nonconvex constraints.
            number_of_constraints = length(obj.constraints);
        end
        function value =  evaluate_coordinate_state_cost(obj,coordinates_state)
            value = 0;
            % if there actually are constraints
            if (obj.get_number_of_constraints() ~= 0)
                value=1;
                for i=1:obj.get_number_of_constraints()
                    h = obj.constraints{i};
                    value = value*nmpccodegen.controller.obstacles.Obstacle.trim( ...
                        h(coordinates_state)...
                        );
                end
            end
        end
    end
end

File: old_code/src_python/nmpccodegen/controller/obstacles/nonconvex_constraints.py
from .obstacle import  Obstacle
import casadi as cd

class Nonconvex_constraints(Obstacle):
    def __init__(self,model):
        """ construct obstacle """
        super(Nonconvex_constraints, self).__init__(model)
        self._constraints=[]
    def add_constraint(self,constraint):
        self._constraints.append(constraint)
    def evaluate_coordinate_state_cost(self,coordinates_state):
        """ evaluate the function h(x) """
        if self.number_of_constraints == 0:
            return 0
        # if there actually are constraints:
        value=1.
        for i in range(0,self.number_of_constraints):
            h = self._constraints[i]
            value *= Obstacle.trim_and_square( \
                h(coordinates_state)\
                )
        return value

    @property
    def number_of_constraints(self):
        return len(self._constraints)